### DISCUSSION TITLE
hh2 temperature shift

### DISCUSSION THREAD

**Discussion by katduecker** [jonescompneurolab/hnn-core#930] [General]
Hi all,

I'm just comparing our hh2.mod to NEURON's hh.mod https://github.com/neuronsimulator/nrn/blob/master/src/nrnoc/hh.mod

q10 in hh.mod is calculated based on the difference between the temperature for the stimulation (in our case 37 degC) and the temperature of the squid giant axon (6.3 degC).

In hh2.mod, there is an additional tshift = 30.7 = 37 - 6.3 that is subtracted.
As such, it seems that q10 is 0 in our mod file.

I noticed that the AP is a lot faster when using the hh.mod mechanism in neuron compared to our hh2. 
Was tshift added to prolong the AP to > 1 ms?

If I'm not mistaken, this seems to be the only difference between them so I'm curious.

Thanks!

**Comment by jasmainak** [comment]
I think @stephanie-r-jones and @samnemo are the appropriate people to answer your question!

**Reply by katduecker** [reply]
Thanks, will do!

**Comment by rythorpe** [comment]
Hey @katduecker, I believe the discrepancy you noticed is designed to maintain the same functionality. We set our temperature to [37.0 C](https://github.com/jonescompneurolab/hnn-core/blob/c291a2271aa89c2ff810027dc72582401882bcb6/hnn_core/network_builder.py#L58), whereas the default hh.mod file assumes the temp is set to [6.3 C](https://github.com/neuronsimulator/nrn/blob/d3a37582b886d00ef21f2853d3bf1a3c038e23ff/src/nrnoc/hh.mod#L12). The `q10` variable should be zero in order to provide consistent dynamics across .mod files, which we accomplish by subtracting by an additional offset: `celsius - temp - tshift == 37.0 - 6.3 - 30.7 == 0`.

I've been concerned by the prolonged AP time as well, which I believe is due to other active mechanisms added to the model, such as CA2+, which is particularly exaggerated in the L5 pyramidal cells. I'm happy to be proven wrong here, but my recollection is that the hh.mod file distributed with NRN has the same sort of dynamics when substituted into one of our cells (try a basket cell since they don't have CA2+ mechs in them and be sure to adjust the temp accordingly).

**Reply by katduecker** [reply]
Hi @rythorpe thanks for this! I have been investigating our hh2 quite thoroughly both in the HNN cells and now by adding the mechanism to a ball and stick cell in NEURON. The long time course definitely only depends on hh2. 

I believe that we do set h.celsius = 37 at the start of the simulation, don't we? The Neuron hh mod file also calculates q10 but the spiking looks different. 

This is a NEURON hh spike with h.celsius = 37. (orange line is just a passive dendrite)

![hh_spike](https://github.com/user-attachments/assets/34a16c4a-bf11-4186-b8f6-1ceb87b7536d)

And this a spike for a NEURON cell to whose soma I have just added hh2. It doesn't  change whether I change celsius or not

![hnn_hh2_spike](https://github.com/user-attachments/assets/1ca8e6df-70b1-4d68-92de-815f9b2d0938)


I do think that spikes in L5 pyramidal cells in human cortex are longer than the 1ms that HH comes up with, they last about 2-4 ms 

https://www-sciencedirect-com.revproxy.brown.edu/science/article/pii/S0896627321006279

Still, HNN cells have a very long refractory period which seems to get me in trouble when I want them spike at ~100 Hz

**Reply by rythorpe** [reply]
@katduecker yeah, the refractory period on these spikes is outrageously long for studying realistic spiking dynamics in the HNN network. When you say that the spike doesn't change whether you change `h.celsius` or not, how are you setting it? Have you tried recompiling a new version of hh2.mod without tshift and seeing if the spikes are faster in HNN-core? The main other variable I can think of that we may need to adjust for is the resting membrane potential, which should be at -65mV. LMK if you want to discuss this in real time - we might be able to get a modified version of the hh2.mod file working for you fairly quickly.

**Comment by samnemo** [comment]
hh2.mod was set to have the same dynamics as hh.mod at 6.3 C with temp set to 37 C

NEURON default temperature is 6.3 C for squid where HH data comes from

When I checked a while ago the original cortical model did not set temperature to 37 C

Compensating for temperature properly  may require other (possibly minor) adjustments to the model to preserve overall dynamics
________________________________

**Reply by rythorpe** [reply]
Thanks for the insight @samnemo. We do set the temp in HNN-core, but perhaps we should double check what the actual value is we're setting it too. In the past, I've adjusted the hh2.mod file to speed up the spikes a little by fiddling with the `gnabar`, `gkbar`, `gl`, and `el` params s.t. the -65mV resting membrane potential assumption is always maintained. If we were to remove `tshift` from the equation, does it sound reasonable to just make sure it doesn't mess with the resting membrane potential (which I don't think it will, but we can make adjustments accordingly if needed) and then proceed with using the updated mod file?

**Reply by samnemo** [reply]
I'd suggest the following long-term plan: restore/use the original hh.mod, test HNN dynamics, if they differ substantially from previous versions then revise the other circuit parameters, such as synaptic weights, to compensate
Short term plan: some of what you suggest could work - resting membrane potential is one possible issue, but overall cellular excitability is another (which may require adjusting synaptic weights)

**Reply by katduecker** [reply]
Hi @samnemo Thanks for the insight! After some digging @asoplata found that hh.mod (which is really our hh2.mod without temperature shift) was fit to cold squid axon. So I guess even reverting to hh.mod and changing temperature might not solve all of our problems? My plot above suggest that hh.mod with celsius=37 produces a spike that is too fast. Since we now want to explore spiking, backpropagating APs, and calcium spikes, it might be worth doing a bit more digging to get this right. We're now planning on gathering current activation/inactivation functions developed for models of human L5 pyramidal neurons. If we find one that looks good, we'll use it off the shelf. If we don't find a good model but find data, we'll try and fit our own. Some of the other channels seem to need updating too, but we'll start with a new hh.mod. At least this is the plan that we want to present to @stephanie-r-jones this week.

**Reply by rythorpe** [reply]
What do you mean by "activation/inactivation functions"? Are you specifically referring to a tuned model with activate calcium dynamics? I anticipate we'll need to re-tune the network parameters regardless....

**Reply by asoplata** [reply]
Hey y'all, firstly thanks for building HNN in the first place! Secondly, I think I understand what Sam meant above by "hh2.mod was set to have the same dynamics as hh.mod at 6.3 C with temp set to 37 C". I did some archeology, and I'm making a guess that this is what happened:

1. In the original Jones 2009 code (github mirror at https://github.com/ModelDBRepository/136803 ), the default NEURON temperature was used. In other words, "celsius" was never changed. Notably, *the NEURON-default 'hh' model mechanisms* were used (see https://github.com/ModelDBRepository/136803/blob/master/sj10-cortex.hoc#L197 ).
2. For the default model used by HNN (original), you wanted to adjust the global temperature of the simulation to something more reasonable like 37 degrees Celsius (which IIRC would slightly affect some values automatically, like Nernst equilibria).
3. However, you wanted to keep the speed of the dynamics of your channels mostly the same as Jones 2009. So, you made a copy of NEURON's `hh.mod` and named it `hh2.mod` (here https://github.com/jonescompneurolab/hnn/blob/0.0.1/mod/hh2.mod#L93 ), with the only real difference being the addition of `tshift`. This `tshift` addition was also made to the channels `ca.mod`, `kca.mod`, and `km.mod` in https://github.com/jonescompneurolab/hnn/tree/0.0.1/mod .
4.  NEURON's `hh.mod`'s default values are based on experimental data for a squid giant axon at 6.3 C. By adding your `tshift`, this changes the mechanisms such that when the simulation is given the global temperature of 37 C, the speed and dynamics of the `hh2` Na and K channels (among other channels) **behave as if the temperature is actually 6.3 C**. 

Is this accurate? I will note that the channels found in https://github.com/jonescompneurolab/hnn/tree/0.0.1/mod  are identical to the current versions in HNN-Core. *edit: a word

**Reply by katduecker** [reply]
> What do you mean by "activation/inactivation functions"? Are you specifically referring to a tuned model with activate calcium dynamics? I anticipate we'll need to re-tune the network parameters regardless....

@rythorpe Nope, we mean Na/K/L in hh2.mod! Calcium will be a different beast

**Reply by samnemo** [reply]
> Hi @samnemo Thanks for the insight! After some digging @asoplata found that hh.mod (which is really our hh2.mod without temperature shift) was fit to cold squid axon. So I guess even reverting to hh.mod and changing temperature might not solve all of our problems? My plot above suggest that hh.mod with celsius=37 produces a spike that is too fast. Since we now want to explore spiking, backpropagating APs, and calcium spikes, it might be worth doing a bit more digging to get this right. We're now planning on gathering current activation/inactivation functions developed for models of human L5 pyramidal neurons. If we find one that looks good, we'll use it off the shelf. If we don't find a good model but find data, we'll try and fit our own. Some of the other channels seem to need updating too, but we'll start with a new hh.mod. At least this is the plan that we want to present to @stephanie-r-jones this week.

temperature set in HNN should be 37 C, but yeah, reverting to hh.mod with 37C temp would still require adjustments since then the channel dynamics will differ due to removal of tshift - so parts of the model will likely need tweaking -->> plan to gather activation/inactivation functions from human data sounds like a good idea

**Reply by samnemo** [reply]
> Hey y'all, firstly thanks for building HNN in the first place! Secondly, I think I understand what Sam meant above by "hh2.mod was set to have the same dynamics as hh.mod at 6.3 C with temp set to 37 C". I did some archeology, and I'm making a guess that this is what happened:
> 
> 1. In the original Jones 2009 code (github mirror at https://github.com/ModelDBRepository/136803 ), the default NEURON temperature was used. In other words, "celsius" was never changed. Notably, _the NEURON-default 'hh' model mechanisms_ were used (see https://github.com/ModelDBRepository/136803/blob/master/sj10-cortex.hoc#L197 ).
> 2. For the default model used by HNN (original), you wanted to adjust the global temperature of the simulation to something more reasonable like 37 degrees Celsius (which IIRC would slightly affect some values automatically, like Nernst equilibria).
> 3. However, you wanted to keep the speed of the dynamics of your channels mostly the same as Jones 2009. So, you made a copy of NEURON's `hh.mod` and named it `hh2.mod` (here https://github.com/jonescompneurolab/hnn/blob/0.0.1/mod/hh2.mod#L93 ), with the only real difference being the addition of `tshift`. This `tshift` addition was also made to the channels `ca.mod`, `kca.mod`, and `km.mod` in https://github.com/jonescompneurolab/hnn/tree/0.0.1/mod .
> 4. NEURON's `hh.mod`'s default values are based on experimental data for a squid giant axon at 6.3 C. By adding your `tshift`, this changes the mechanisms such that when the simulation is given the global temperature of 37 C, the speed and dynamics of the `hh2` Na and K channels (among other channels) **behave as if the temperature is actually 6.3 C**.
> 
> Is this accurate? I will note that the channels found in https://github.com/jonescompneurolab/hnn/tree/0.0.1/mod are identical to the current versions in HNN-Core. *edit: a word

yes, the above is accurate ... good job digging through the code/history

**Comment by samnemo** [comment]
The hh model is a decent approximation of the “undoing” shift in hh2.mod is removed and temp is set properly

But then the full HNN model will probably need some other adjustments to compensate for altered neuronal dynamics … all doable and probably not major effort required
________________________________

**Reply by katduecker** [reply]
hh.mod is the same as hh2.mod but without the tshift if I'm seeing that correctly?

**Reply by asoplata** [reply]
I believe you're exactly correct @katduecker . Here is a diff between the current NEURON `hh.mod` on the left and the `hh2.mod` that HNN has always used on the right. The only mechanical difference is the temperature adjustment.
![Screenshot_20241114_101941](https://github.com/user-attachments/assets/ed14a2a8-1163-4102-859e-57d400e5942d)

**Comment by samnemo** [comment]
by the way, once going back to single cell dynamics/parameter adjustments, might also consider fitting firing rate vs current injection for some of the cell types in the model

**Reply by katduecker** [reply]
Thanks, that's a good one for the objective functions!
